www/blog/2024-05-23-trpc-actions.mdx

```diff
@@ -0,0 +1,265 @@
+Sidenote, you can even use [tRPC directly to do this](https://github.com/mmkal/trpc-cli). But that's not what we're here to talk about today,
+we're going to talk about how to use tRPC with server actions from Next.js.
+
+## What's a server action?
+
+In case you live under a rock and haven't kept up with the latest React and Next.js features, server actions allows you to write regular functions that are executed on the server, import them on the client and call them just as if they were regular functions.
+You may think that this sounds similar to tRPC which is true. According to Dan Abramov, server actions are tRPC as a bundler feature:
+
+<div className="mb-4 flex w-full justify-center">
+  <Tweet id="1756120297529368801" />
+</div>
+
+And this is totally accurate, server actions are, similar to tRPC, at the end of the day an RPC technology. Both allows you to write functions on the backend and call them with full typesafety on the frontend with the network layer abstracted away.
+
+So where does tRPC come in? Why do I need both tRPC and server actions? Server actions is a primitive, and as for all primitives they're quite barebones and thus lacking some fundamental aspects when it comes to building APIs. For any API endpoint that is exposed over internet, you need to validate and authorize requests to ensure your API is not maliciously used. As previously mentioned, tRPC's API is appreciated by the community, so wouldn't it be nice if we could use tRPC to define server actions and utilizing all the awesome features that comes built-in with tRPC such as input validation, authentication and authorization through middlewares, output validation, data transformers, etc, etc. I think so, so let's dig in.
+
+## Defining server actions with tRPC
+
+:::note
+**Prerequisites:** In order to use server actions, you need to use the Next.js App Router. Additionally, all the tRPC stuff we'll use are only available on tRPC v11, so make sure you're using the beta release channel of tRPC:
+
+<InstallSnippet pkgs="@trpc/server@beta" />
+:::
+
+Let's start off by initializing tRPC and defining our base server actions procedure.
+We'll use the `experimental_caller` method on the procedure builder, which is a new method that allows you to
+customize the way that the procedure is called when it's invoked as a function. We'll also use the adapter `experimental_nextAppDirCaller`
+to make it compatible with Next.js. This adapter will handle cases where the server action is wrapped in `useActionState` on the client,
+which [changes the call signature of the server action](https://react.dev/reference/react/useActionState#my-action-can-no-longer-read-the-submitted-form-data).
+
+We'll also be using a `span` property as [metadata](/docs/server/metadata), since there is no ordinary path like when you use a router (`user.byId` for example). You can use the span property to differentiate procedures, for example during logging or observability.
```

Comment:
is `span` a normal word to use in observability? Otherwise, I'd just call it `name` or `path`.

Span threw me off a bit when reading as I only see it as a HTML-thing

Reply:
yea span is from the observability world